File: tui/src/scala/tui/internal/debug_assert.scala
package tui
package internal

object debug_assert {
  def apply(pred: Boolean, msg: String, details: Any*): Unit =
    if (pred) ()
    else {
      // todo: template in at correct place
      val formattedDetails = if (details.isEmpty) "" else details.mkString(" (", ", ", ")")
      sys.error(s"assertion failed: $msg$formattedDetails")
    }
}

File: tui/src/java/tui/internal/DebugAssert.java
package tui.internal;

public final class DebugAssert {
  private DebugAssert() {}

  public static void apply(boolean pred, String msg, Object... details) {
    if (!pred) {
      String formatted;
      if (details.length == 0) {
        formatted = "";
      } else {
        StringBuilder sb = new StringBuilder(" (");
        for (int i = 0; i < details.length; i++) {
          if (i > 0) sb.append(", ");
          sb.append(details[i]);
        }
        sb.append(")");
        formatted = sb.toString();
      }
      throw new RuntimeException("assertion failed: " + msg + formatted);
    }
  }
}
